### PR TITLE
Add dry run functionality for pause and unpause commands

### PR DESCRIPTION
**Pull request by fossedihelm**
**What this PR does / why we need it**:
Allows user to use --dry-run flag for pause and unpause virtctl commands.
Passing this flag to one of these command will do all checks showing possible failures, but changes will not be effective.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
added DryRun mode to virtcl for pause and unpause commands
pause and unpause requests now have a body with PauseOptions and UnpauseOptions, respectively
```

### CODE REVIEW

File: pkg/virt-api/rest/subresource.go [hunk 1]
```diff
@@ -107,17 +107,19 @@ func (app *SubresourceAPIApp) fetchAndValidateVirtualMachineInstance(namespace,
...
 	if statusErr != nil {
 		writeError(statusErr, response)
 		return
 	}
 
-	err := conn.Put(url, app.handlerTLSConfiguration, request.Request.Body)
-	if err != nil {
-		writeError(errors.NewInternalError(err), response)
-		return
+	if !dryRun {
```
Comment (jean-edouard) [inline pkg/virt-api/rest/subresource.go:117]: How about `if dryRun { return }`? (smaller diff, less indent..)

File: pkg/virt-api/rest/subresource_test.go [hunk 2]
```diff
@@ -1735,70 +1735,85 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
...
+			request.Request.Body = io.NopCloser(bytes.NewReader(bytesRepresentation))
 
 			app.UnpauseVMIRequestHandler(request, response)
 
 			Expect(response.StatusCode()).To(Equal(http.StatusOK))
-		})
+
+		},
+			table.Entry("", true, true, &v1.UnpauseOptions{}),
+			table.Entry("with dry-run option", true, true, &v1.UnpauseOptions{DryRun: []string{k8smetav1.DryRunAll}}),
```
Comment (jean-edouard) [inline pkg/virt-api/rest/subresource_test.go:1815]: The resulting sentence for this test will be "Should unpause a running, paused VMI with dry-run option".
That sounds incorrect, nothing should actually happen in dry mode...

File: pkg/virtctl/pause/pause.go [hunk 3]
```diff
@@ -61,6 +65,7 @@ Second argument is the name of the resource.`,
 			return c.Run(args)
 		},
 	}
+	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "--dry-run=false: Flag used to set whether perform a dry run or not. If true the command will executed without performing any changes.")
```
Comment (jean-edouard) [inline pkg/virtctl/pause/pause.go:68]: Missing `to` after `whether`, like in the other PR :)
Also, "will **be** executed" maybe?

File: pkg/virtctl/pause/pause.go [hunk 4]
```diff
@@ -82,6 +87,7 @@ Second argument is the name of the resource.`,
...
```
Comment (jean-edouard) [inline pkg/virtctl/pause/pause.go:90]: Same as above

**Review by jean-edouard — COMMENTED** [review]
Few in-line comments.
If https://github.com/kubevirt/kubevirt/pull/6681 merges soon, maybe my nits there can also be addressed here?

File: pkg/virtctl/pause/pause.go (hunk 3 above)
Comment (fossedihelm) [reply pkg/virtctl/pause/pause.go:68]: I'm really tenacious doing the same mistake :) Thanks, will fix

File: pkg/virt-api/rest/subresource.go (hunk 1 above)
Comment (fossedihelm) [reply pkg/virt-api/rest/subresource.go:117]: Thanks, sounds better :)

Comment (fossedihelm) [reply pkg/virt-api/rest/subresource.go:117]: Anyway, Yes, I'm fixing your nits in [#6681](https://github.com/kubevirt/kubevirt/pull/6681) in this PR

File: pkg/virtctl/pause/pause.go (hunk 4 above)
Comment (fossedihelm) [reply pkg/virtctl/pause/pause.go]: Thanks. Fixed

File: pkg/virtctl/pause/pause.go (hunk 3 above)
Comment (fossedihelm) [reply pkg/virtctl/pause/pause.go]: Thanks again for the review. Fixed

File: pkg/virt-api/rest/subresource_test.go (hunk 2 above)
Comment (fossedihelm) [reply pkg/virt-api/rest/subresource_test.go:1837]: Thanks for the review and suggestions. Fixed

File: pkg/virt-api/rest/subresource.go (hunk 1 above)
Comment (fossedihelm) [reply pkg/virt-api/rest/subresource.go]: Fixed this and [#6681](https://github.com/kubevirt/kubevirt/pull/6681) nits. Thanks

File: pkg/virt-api/rest/subresource_test.go (hunk 2 above)
Comment (jean-edouard) [reply pkg/virt-api/rest/subresource_test.go:1837]: Same as above, it doesn't seem needed to add the booleans as parameters.

In general, having random booleans around is hard to read and usually frowned upon.
How would you feel about defining 4 consts at the top of the file like "Paused=true, UnPaused=false, Running=true, NotRunning=false"?
That way we could write calls like `expectVMI(Running, Paused)`, without having to create function parameters for them!

**Review by jean-edouard — COMMENTED** [review]
Looks great overall, just a couple comments in-line

File: pkg/virt-api/rest/subresource_test.go [hunk 5]
```diff
@@ -1757,70 +1757,85 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
...
-		})
-
-		It("Should fail pausing a not running VMI", func() {
 
-			expectVMI(false, false)
-
-			app.PauseVMIRequestHandler(request, response)
+		},
+			table.Entry("with default", true, false, &v1.PauseOptions{}),
+			table.Entry("with dry-run option", true, false, &v1.PauseOptions{DryRun: []string{k8smetav1.DryRunAll}}),
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-api/rest/subresource_test.go]: You're right. Missed decoupling two tests. Thanks

File: pkg/virt-api/rest/subresource_test.go (hunk 2 above)
Comment (fossedihelm) [reply pkg/virt-api/rest/subresource_test.go]: I like it! It will increase readability. Thanks so much!

**Comment by fossedihelm** [comment]
@jean-edouard Fixed and rebased due to conflicts

File: pkg/virt-api/rest/subresource_test.go [hunk 6]
```diff
@@ -53,7 +53,13 @@ import (
 	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
 )
 
-const vmPathFormat = "/apis/kubevirt.io/%s/namespaces/%s/virtualmachines/%s"
+const (
+	vmPathFormat = "/apis/kubevirt.io/%s/namespaces/%s/virtualmachines/%s"
+	Running      = true
```
Comment (xpivarc) [inline pkg/virt-api/rest/subresource_test.go:58]: :+1:

File: pkg/virtctl/pause/pause.go [hunk 7]
```diff
@@ -110,6 +116,11 @@ func (vc *VirtCommand) Run(args []string) error {
 		return fmt.Errorf("Cannot obtain KubeVirt client: %v", err)
 	}
 
+	var dryRunOption []string
+	if dryRun {
+		fmt.Printf("Dry Run execution\n")
```
Comment (xpivarc) [inline pkg/virtctl/pause/pause.go:121]: `fmt.Println`

File: tests/pausing_test.go [hunk 8]
```diff
@@ -366,6 +410,49 @@ var _ = Describe("[rfe_id:3064][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
+				Consistently(func() bool {
+					updatedVm, err := virtClient.VirtualMachine(util.NamespaceTestDefault).Get(vm.Name, &v12.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					for _, condition := range updatedVm.Status.Conditions {
+						if condition.Type == v1.VirtualMachinePaused && condition.Status == k8sv1.ConditionTrue {
+							return false
+						}
+					}
+					return true
+				}, time.Duration(5)*time.Second, 5).Should(BeTrue())
```
Comment (xpivarc) [inline tests/pausing_test.go:429]: `5*time.Second` is enough.

File: staging/src/kubevirt.io/api/core/v1/types.go [hunk 9]
```diff
@@ -1734,6 +1734,38 @@ type StartOptions struct {
...
+	// request. Valid values are:
+	// - All: all dry run stages will be processed
+	// +optional
+	// +listType=atomic
+	DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`
+}
+
+// UnpauseOptions may be provided on unpause request.
+//
+// +k8s:openapi-gen=true
```
Comment (xpivarc) [inline staging/src/kubevirt.io/api/core/v1/types.go:1755]: Is this necessary?

**Review by xpivarc — APPROVED** [review]
Looks good. Nice improvements on the way.
One question only. Is the OpenAPI necessary? 
/hold

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 9 above)
Comment (fossedihelm) [reply staging/src/kubevirt.io/api/core/v1/types.go:1755]: Thanks. I think it is not necessary, probably a mistake too. I will fix it

File: tests/pausing_test.go (hunk 8 above)
Comment (fossedihelm) [reply tests/pausing_test.go]: Thanks for the advice :slightly_smiling_face:

**Comment by fossedihelm** [comment]
> Looks good. Nice improvements on the way. One question only. Is the OpenAPI necessary? Everything else can be ignored. /hold

Thanks for the review, the OpenAPI was not necessary. I pushed the changes

**Pull request merged**